### PR TITLE
Add heredoc syntax support for RUN/COPY/ADD

### LINKED ISSUE #245: RUN/COPY/ADD heredoc syntax not supported; treated as shell-form literal
## Summary

BuildKit supports heredoc syntax in `RUN`, `COPY`, and `ADD` instructions. The C# parser does not support heredoc syntax and treats `<<EOF` and subsequent lines as part of the shell-form command literal.

## Example

Input: `RUN <<EOF` followed by `echo hello` and `EOF` on separate lines.

C# output: The entire heredoc (including the `<<EOF` marker, the body, and the closing delimiter) is collapsed into one large `LiteralToken` as if it were shell-form command text.

Expected (BuildKit/Lean): Should produce a `heredoc` aggregate token with structured child tokens for the delimiter, body content lines, and closing delimiter.

## Affected Instructions

- `RUN <<DELIM`
- `COPY <<DELIM /path`
- `ADD <<DELIM /path`

## Impact

Any Dockerfile using heredoc syntax will be silently misparsed. The token structure will be wrong.

## Root Cause

The C# parser has no heredoc parser combinator. The `ParseHelper` does not recognize `<<` as a heredoc redirection operator. Adding heredoc support requires implementing a two-pass parser: detecting the opening `<<DELIM` during instruction parsing, then consuming the body until the closing delimiter.

### PR DESCRIPTION
**Pull request by mthalman**
## Summary
- Implement heredoc syntax parsing (`<<DELIM`, `<<-DELIM`, `<<"DELIM"`, `<<'DELIM'`) for RUN, COPY, and ADD instructions
- New `HeredocToken` aggregate token representing the complete heredoc block (marker, body lines, closing delimiter)
- `DockerfileParser` updated to detect heredoc markers during line assembly and consume body lines until closing delimiter
- Custom Sprache parser (`ParseHelper.Heredoc()`) handles context-dependent delimiter matching
- `RunInstruction.Command` is now nullable (returns null for heredoc instructions); `Heredocs` property added
- `FileTransferInstruction.Heredocs` property added for COPY/ADD heredoc access

Fixes #245

## Test plan
- [x] All 693 existing tests pass unchanged
- [x] Heredoc parsing for RUN instruction (simple, chomp flag, quoted delimiters, multiple body lines, empty body, custom delimiters, mount flags, CRLF)
- [x] Heredoc parsing for COPY instruction (simple, quoted delimiter, multiple body lines, with flags)
- [x] Heredoc parsing for ADD instruction (simple, quoted delimiter, multiple body lines)
- [x] Quoted and unquoted delimiters (double-quoted, single-quoted, chomp+quoted)
- [x] Round-trip fidelity verified for all heredoc forms (Parse(text).ToString() == text)
- [x] Dockerfile-level round-trip with heredocs between other instructions
- [x] Token structure validation (child tokens of HeredocToken)
- [x] Edge cases: empty lines, special characters, no trailing newline, delimiter with numbers/underscores
- [x] DockerfileParser.ExtractHeredocDelimiters unit tests
- [x] 45 new tests total, 738 tests pass

### CODE REVIEW

File: src/Valleysoft.DockerfileModel/ParseHelper.cs [hunk 1]
```diff
@@ -982,6 +982,171 @@ public TokenWrapper(string openingString, string closingString)
...
+        while (!current.AtEnd && current.Current != '\n' && current.Current != '\r')
+        {
+            restOfLineChars.Add(current.Current);
+            current = current.Advance();
+        }
+
+        if (restOfLineChars.Count > 0)
+        {
+            tokens.Add(new StringToken(new string(restOfLineChars.ToArray())));
+        }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel/ParseHelper.cs]: Addressed in 3ca424f: added a detailed known-limitation comment at the rest-of-marker-line consumption site in HeredocParseImpl explaining why the destination and subsequent <<DELIM markers are swallowed, and why the full parser restructuring is deferred.

File: src/Valleysoft.DockerfileModel/DockerfileParser.cs [hunk 2]
```diff
@@ -87,6 +140,27 @@ public static Dockerfile ParseContent(string text)
...
+            // Group 1 = chomp flag '-' (may be empty), Group 2 = double-quoted,
+            // Group 3 = single-quoted, Group 4 = unquoted
+            bool hasChomp = match.Groups[1].Value == "-";
+            string delimiter = match.Groups[2].Success ? match.Groups[2].Value :
+                               match.Groups[3].Success ? match.Groups[3].Value :
+                               match.Groups[4].Value;
+            delimiters.Add((delimiter, hasChomp));
+        }
+        return delimiters;
+    }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel/DockerfileParser.cs]: Fixed in e2b1082 (regex change) + 44a03df (tests): tightened the quoted-delimiter capture groups in HeredocMarkerRegex from [^"]+/[^']+ to [A-Za-z0-9_.\-]+, matching the expanded IsHeredocDelimiterChar. Tests verify that quoted delimiters with hyphens round-trip correctly and that characters outside the accepted set are not matched.

File: src/Valleysoft.DockerfileModel.Tests/HeredocTests.cs [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+
+    [Fact]
+    public void Copy_HeredocWithFromFlag_RoundTrips()
+    {
+        // COPY --from with heredoc source
+        string text = "COPY <<EOF /app/script.sh\necho hello\nEOF\n";
+        CopyInstruction result = CopyInstruction.Parse(text);
+        Assert.Equal(text, result.ToString());
+        Assert.Single(result.Heredocs);
+    }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel.Tests/HeredocTests.cs]: Fixed in 822610b: renamed Copy_HeredocWithFromFlag_RoundTrips to Copy_HeredocWithDestination_RoundTrips and updated the inline comment to accurately describe what the test validates (a heredoc COPY with a destination path, no --from flag).
